southworks/BotBuilder-Samples
# [FOR REVIEW] Visual Studio Extension for adding Dialogs Template from context menu

## Proposed Changes
Add a Visual Studio Extension project for adding a Dialogs Template to a bot project.
The extension consists of an Item Template that allows the users adding the classes and accessors needed to use Dialogs into a .NET Bot Project. 
The tool will make the following changes to the project:

1. Install NuGet package Microsoft.Bot.Builder.Dialogs v 4.2.2
2. Add the following classes:
- Class with the name the user decides that inherits from ComponentDialog class and implements a waterfall dialog for greeting the user.
- GreetingState class for managing the user state properties for Greeting.
3. Update the following classes:
- Startup class: it will add a Memory Storage and save the conversation state
- Bot class: it will add the accessors properties and modify the constructor to create a DialogSet. Also it will begin or continue the conversation according to the dialog context. 

### Future features:
* Support any bot project. This version works only with an **EmptyBot** project. 
* Show the user a wizard to personalize the dialog control flow.

Here are some Screenshots:

The user creates an **EmptyBot** project:
![image](https://user-images.githubusercontent.com/44245136/53249972-2c705900-3698-11e9-971b-d874be7ac0cb.png)

Right-click on the project to Add a New Item:
![image](https://user-images.githubusercontent.com/44245136/53249866-eb784480-3697-11e9-87f4-18069d3a31dd.png)

Select Dialogs Template Item:
![image](https://user-images.githubusercontent.com/44245136/53250095-7b1df300-3698-11e9-8d6b-d88e18a43580.png)

Accept the changes clicking on the `Discard` button:
![image](https://user-images.githubusercontent.com/44245136/53250197-ab659180-3698-11e9-872c-0b41fe71a3c6.png)

The solution explorer before and after adding the Dialogs:
![image](https://user-images.githubusercontent.com/44245136/53250552-62620d00-3699-11e9-9f94-1398a663e34e.png)


## Testing
First, go to `{your repository directory}/generators/vsix-vs-win/Dialogs Template` and open the `DialogsTemplate.sln` solution
Then, there are two alternatives

_Alternative A_

* Build the VSIX project
* Go to `bin/debug` folder under your project's directory and run the `DialogsTemplateVSIX.vsix` file to install the extension in your Visual Studio (**NOTE**: first, ensure your Visual Studio instance is closed)
* Once the extension is installed, open Visual Studio and create a new **EmptyBot** project.
* After the project is loaded, _right-click_ on it and select _Add_ -> _New Item_ in the context menu. Then select _Dialogs Template_ item and name it as you like.

_Alternative B_

* Build the VSIX project
* Run the project. It will open a Visual Studio Experimental Instance
* In the Experimental instance, open an **EmptyBot** project.
* After the project is loaded, _right-click_ on it and select _Add_ -> _New Item_ in the context menu. Then select _Dialogs Template_ item and name it as you like.

This PR is related to the [ISSUE #176](https://github.com/southworkscom/BotBuilder-Samples/issues/176)
